### PR TITLE
WorldEdit commands for clipboard/schematics

### LINKED ISSUE #262: WorldEdit Clipboard Support
WorldEdit has a clipboard for each player. It would be nice to be able to place a schematic/cuboid into a player's clipboard.
I've done something like this in the past, but that was done using only FAWE's implementation.
The command could be like:
`- fawe clipboard target:<player> filePath:schematics/<def[title]>`
`- fawe clipboard target:<def[value]> cuboid:<def[cuboid]> position:<def[location]>`

The reason that clipboard is an arg and not a command on its own is to allow additional suppot for FAWE such as editing blocks (//walls //sphere //pyramid etc.) or doing a massive schematic paste with undo supported.

### PR DESCRIPTION
**Pull request by Mwthorn**
Fixes #262 

Tested on Paper version 1.13.2

### CODE REVIEW

File: src/main/java/com/denizenscript/depenizen/bukkit/bridges/WorldEditBridge.java [hunk 1]
```diff
@@ -12,6 +14,8 @@
     @Override
     public void init() {
         instance = this;
+        DenizenAPI.getCurrentInstance().getCommandRegistry().registerCoreMember(WorldEditCommand.class,
+                "WORLDEDIT", "worldedit [cuboid_to_schematic/cuboid_to_clipboard/schematic_to_paste/schematic_to_clipboard] (rotate:<amount>) (target:<player>)", 2);
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/bridges/WorldEditBridge.java]: Doc line is wrong

File: src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+import org.bukkit.Bukkit;
+import org.bukkit.entity.Player;
+
+import java.io.*;
+import java.net.URLDecoder;
+
+public class WorldEditCommand extends AbstractCommand {
+    // <--[command]
+    // @Name worldedit
+    // @Syntax worldedit [CUBOIDSCHEMATIC/CUBOIDCLIPBOARD/SCHEMATICPASTE/SCHEMATICCLIPBOARD] (file:<filepath>) (undoable) (noair) (cuboid:<CuboidTag>) (rotate:<amount>) (target:<playerTag>)
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: what even is this doc line? Please conform it to the standards set by every existing command

File: src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+                scriptEntry.addObject("action", arg.asElement());
+            }
+
+            else {
+                arg.reportUnhandled();
+            }
+
+        }
+        if (!scriptEntry.hasObject("action")) {
+            throw new InvalidArgumentsException("Action not specified! (play/stop)");
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: Nonsense copypasta

File: src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+        }
+        if (!scriptEntry.hasObject("target")) {
+            if (Utilities.entryHasPlayer(scriptEntry)) {
+                scriptEntry.addObject("target", Utilities.getEntryPlayer(scriptEntry));
+            }
+            else {
+                throw new InvalidArgumentsException("Must specify players to add, remove or spectate!");
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: Nonsense copypasta

File: src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+                try (EditSession editSession = WorldEdit.getInstance().getEditSessionFactory().getEditSession(w, -1,null,p)) {
+                    Operation operation = holder
+                            .createPaste(editSession)
+                            .to(BlockVector3.at(position.getBlockX(), position.getBlockY(), position.getBlockZ()))
+                            .ignoreAirBlocks(pastenoair)
+                            .build();
+                    Operations.complete(operation);
+                    editSession.flushSession();
+                    WorldEdit.getInstance().getSessionManager().get(p).remember(editSession);
+                } catch (WorldEditException e) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: Please conform to basic Denizen formatting standards

Comment (Mwthorn) [reply src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: I'm not sure what you mean here. Is it the inside of the `try` clause or do I need to change `WorldEditException` to `Exception` ?

Comment (mcmonkey4eva) [reply src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: Basic formatting of a try/catch

File: src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+    // @Syntax worldedit [cuboidschematic/cuboidclipboard/schematicpaste/schematicclipboard] (file:<file path>) (undoable) (noair) (cuboid:<CuboidTag>) (rotate:<amount>) (target:<player>)
+    // @Group Depenizen
+    // @Plugin Depenizen, WorldEdit
+    // @Required 2
+    // @Short Allows access to various WorldEdit actions.
+
+    // @Description
+    // Allows access to various WorldEdit actions.
+    // Actions can be cuboidschematic, cuboidclipboard, schematicpaste, schematicclipboard
+    // If no targets are specified, the target will be the player in the queue.
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: This line implies the input is a list, which it isn't. If there isn't a list, the input doesn't need to exist tbh (just use `player:` arg)

File: src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+import org.bukkit.Bukkit;
+import org.bukkit.entity.Player;
+
+import java.io.*;
+import java.net.URLDecoder;
+
+public class WorldEditCommand extends AbstractCommand {
+    // <--[command]
+    // @Name worldedit
+    // @Syntax worldedit [cuboidschematic/cuboidclipboard/schematicpaste/schematicclipboard] (file:<file path>) (undoable) (noair) (cuboid:<CuboidTag>) (rotate:<amount>) (target:<player>)
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: Please conform doc line to Denizen standards.
These two arguments specifically: ` (cuboid:<CuboidTag>) (rotate:<amount>)`

File: src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+
+    // @Description
+    // Allows access to various WorldEdit actions.
+    // Actions can be cuboidschematic, cuboidclipboard, schematicpaste, schematicclipboard
+    // If no targets are specified, the target will be the player in the queue.
+    // Specify noair to paste without air blocks from the schematic.
+    // Specify rotate to rotate the schematic when pasting it.
+    // Specify undoable to attach the paste to a player's WorldEdit history which allows them to undo/redo.
+    // The file path starts in the folder: /plugins/Denizen/schematics/
+
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: Please conform to Denizen meta formatting standards (on these whitespace lines)

**Comment by Mwthorn** [comment]
Some days have passed with this pull request.
I just want to be on the same page; is there anything that I missed in your reviews?
I am aware the last few commits are weird but I'm sure that will be fixed in one squashed commit.

**Comment by mcmonkey4eva** [comment]
Please fix the broken commits. I can't review the PR with those like that.

File: src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+            else {
+                arg.reportUnhandled();
+            }
+
+        }
+        if (!scriptEntry.hasObject("action")) {
+            throw new InvalidArgumentsException("Action not specified!");
+        }
+        if (Utilities.entryHasPlayer(scriptEntry)) {
+            scriptEntry.addObject("target", Utilities.getEntryPlayer(scriptEntry));
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: What is this? If it's not user-specifiable, it should not be a script entry object...

File: src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+        ElementTag file = scriptEntry.getObjectTag("file");
+        ElementTag action = scriptEntry.getObjectTag("action");
+        PlayerTag target = scriptEntry.getObjectTag("target");
+        LocationTag position = scriptEntry.getObjectTag("position");
+        ElementTag noAir = scriptEntry.getObjectTag("noair");
+        CuboidTag cuboid = scriptEntry.getObjectTag("cuboid");
+        ElementTag id = scriptEntry.getObjectTag("id");
+        ElementTag data = scriptEntry.getObjectTag("data");
+        ElementTag undoable = scriptEntry.getObjectTag("undoable");
+        ElementTag rotate = scriptEntry.getObjectTag("rotate");
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: Several of the options above (id, data, ...) don't correspond to anything that's ever set in parseArgs.

**Comment by mcmonkey4eva** [comment]
What benefit/use does this PR have over just using the native Denizen `schematic` command?

**Comment by Mwthorn** [comment]
Many plugins have hooked into WorldEdit which either limits or changes the way it places the blocks. The clipboard is used along with other tools in worldedit (not just for simply pasting). A significant difference is using this with FastASyncWorldEdit.

File: src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java (hunk 9 above)
Comment (mcmonkey4eva) [reply src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: Please do not use Depenizen code as a reference. Use Denizen as a reference.

**Comment by mcmonkey4eva** [comment]
@Mwthorn when do you plan to clean this up?

**Comment by Mwthorn** [comment]
Cleaned up and also tested it on 1.14.4 Paper

File: src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+
+        // Report to dB
+        Debug.report(scriptEntry, getName(), action.debug()
+                + (target != null ? target.debug() : "")
+                + (position != null ? position.debug() : "")
+                + (noAir != null ? noAir.debug() : "")
+                + (cuboid != null ? cuboid.debug() : "")
+                + (rotate != null ? rotate.debug() : "")
+                + (file != null ? file.debug() : ""));
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: This is missing values (undoable, ...)

File: src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+import com.sk89q.worldedit.util.io.Closer;
+import com.sk89q.worldedit.world.World;
+
+import java.io.*;
+import java.net.URLDecoder;
+
+public class WorldEditCommand extends AbstractCommand {
+    // <--[command]
+    // @Name worldedit
+    // @Syntax worldedit [cuboidschematic/cuboidclipboard/schematicpaste/schematicclipboard] (file:<file path>) (undoable) (noair) (cuboid:<cuboid>) (rotate:<#>)
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: Position is missing from the syntax line

File: src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+    // @Group Depenizen
+    // @Plugin Depenizen, WorldEdit
+    // @Required 2
+    // @Short Allows access to various WorldEdit actions.
+    //
+    // @Description
+    // Allows access to various WorldEdit actions.
+    // Actions can be cuboidschematic, cuboidclipboard, schematicpaste, schematicclipboard
```
[the remark this answers is not in the document]
Comment (Mwthorn) [reply src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: In cases when both args are specified, should I choose one priority or present a debug error?

[the remark this answers is not in the document]
Comment (mcmonkey4eva) [reply src/main/java/com/denizenscript/depenizen/bukkit/commands/worldedit/WorldEditCommand.java]: Error would be better

**Pull request merged**